### ISSUE TITLE
Vulkan support

### ISSUE THREAD

**Issue by RangerMauve** [huggingface/candle#1810]
The m2 metal support is great for people running macs, however it doesn't cover the majority of devices out there. Supporting the Vulkan graphics library can greatly increase the reach of GPUs that could be used for inference.

On a related note, [llama.cpp recently landed support for it](https://github.com/ggerganov/llama.cpp/commit/2307523d322af762ae06648b29ec5a9eb1c73032) and tools like Ollama are working on integrating this support.

**Comment by oddpxl** [comment]
I'd like to second the wish for that - support for Vulkan in candle would be great.

**Comment by niklasha** [comment]
I don't want to promise too much, but I'll mention it here, just in case anybody else is planning or doing similar work: I have started to add Vulkan support to Candle in my private fork.  It is not far gone, I have not even started doing any kernel work, just initializing buffers, so it's not worth to share just yet.  My main motivation for this is for making Candle work on OpenBSD with Radeon, which is otherwise not supported in any way.

**Comment by andar1an** [comment]
Is there any status or updates on this? Would love to follow progress.

**Comment by niklasha** [comment]
I have been sidetracked since I have got a very pressed schedule this fall, hope to get back to it, but currently I cannot say when.  The branch which I am working in can be found [here](https://github.com/niklasha/candle/tree/niklas-vulkan).  But it is not much to look at right now I am afraid.

**Comment by RobbyCBennett** [comment]
Thank you for working on this.

I use llama.cpp and I have an idea that could help with Vulkan startup speed. I've noticed that llama.cpp works with shaders, but it compiles the shaders each time.

If the future implementation of a Vulkan backend for Candle uses shaders, then it should use shader caching.

**Comment by RobbyCBennett** [comment]
> Do you mean the shader compilation time in your llama.cpp usage is actually of any concern?
Yes, it takes about a second to compile shaders. It's not a crazy amount of time, but it would be nice to cache the shaders on the disk between runs. This is a practice actually used in the emulator Dolphin, and probably in other places too.

This was just an idea for a future improvement. I don't know much about Vulkan, but thanks for your effort.

**Comment by niklasha** [comment]
Just wanted to share that I this night managed to get my first sensible response from a Vulkan-powered candle LLM!

Still rough, unfinished and unoptimized (esp. loading GPU-buffers are slow, but the tag [0.8.3-vulkan-0](https://github.com/niklasha/candle/tree/0.8.3-vulkan-0) in my fork managed to run this on a Macbook Pro M1:

```
cargo run --package candle-examples --example based --features vulkan -- --which 1b-50b --prompt "Flying monkeys are" --sample-len 100
```
resulting in:
```
avx: false, neon: true, simd128: false, f16c: false
temp: 0.00 repeat-penalty: 1.10 repeat-last-n: 64
retrieved the files in 2.164166ms
loaded the model in 26.239473333s
Flying monkeys are a common sight in the wild, but they are not native to the island.

The species is also known as the "flying squirrel" and are found throughout the Americas. They are found from Mexico to Panama and are found in the Caribbean and Central America. The name "flying squirrels".

References 

 

Category:Mammals of Costa Rica, Guatemala, Honduras, and Panama.

External links

Category:Malesia
Category:Mamm
100 tokens generated (1.07 token/s)
```

Interesting enough, the version of candle I am basing this on (I have not yet rebased on current) fails with Metal on my Mac.  And the CPU output is worse (it diverges after 10 tokens or so, probably because of rounding differences in the implementations, the RNG is the same so it is not that) from my perspective (the text has lower coherency IMO).

But here's the real news.  I managed to run it on an OpenBSD system with an AMD RX 6600 XT with 8GB VRAM as well!  37% slower than my 64GB M1, but still!

So a good PoC I think. Now I just wish I had some good AMD GPUs (like 7900 XTX, MI100, W7900 or better :-)  to get really motivated for persuing this fun project in a little more regular way.  Any takers for a donation? :-)

**Comment by niklasha** [comment]
I am proof reading device.rs for Vulkan, and stumble over this in to_cpu:
let mut builder = AutoCommandBufferBuilder::primary(
                Arc::new(StandardCommandBufferAllocator::new(
                    self.device.clone(),
                    StandardCommandBufferAllocatorCreateInfo::default(),
                )),
                self.queue.queue_family_index(),
                CommandBufferUsage::OneTimeSubmit,
            )
            .map_err(VulkanError::ValidatedVulkanError)?;

Why do we use OneTimeSubmit?  to_cpu is called plentiful? is this a residual from the PoC times? or is there a reason ?

**Comment by mateodelnorte** [comment]
Promising work @niklasha. Would this also bring candle support to i7 and i9 Macs with Radeon GPUs? 

Any way others can help support your efforts?

**Comment by niklasha** [comment]
> Promising work [@niklasha](https://github.com/niklasha). Would this also bring candle support to i7 and i9 Macs with Radeon GPUs?

I would expect so, using MoltenVK, but I have no idea unless I get access to such hardware.  Different GPU models supports different features, and may require subtle differences in shader implementations, to either work, or work performantly.  However, I am a bit intrigued, such Macs *should* provide the Metal API, and thus candle *should* already work.  Doesn't it?

> Any way others can help support your efforts?

Right now, I am stalling a bit on vulkano (a Vulkan wrapper for Rust) being in flux with respect to synchronization, where its old model is slightly broken, and won't be fixed, and its new one (known has vulkano-taskgraph) is not quite finished.  

There is of course other stuff to do for keen people, like writing quantization support shaders etc, or figure out and fixmissing operations for various use cases.  But it mostly boils down to getting your hands dirty and dive into the source code.  I don't need testers, it is no way near such stability, and I definitely don't have the bandwidth to deal with a backlog of bug reports.

The only other way to support right now is to motivate me :-)  That would be by supplying better hardware than I have available which basically means larger AMD GPUs in terms of VRAM, at least 16GB, but preferrably 24GB+ cards.

As I said currently I am stalling and spending my spare time elsewhere, only occasionally checking in on the status in vulkano-land and doing some experiments with taskgraph.  But I will come back, I do have incentives to finish this, it's just not top-priority right now.

**Comment by setoelkahfi** [comment]
I'd like to see this! I can only help with testing for now.

**Comment by thewh1teagle** [comment]
Really looking for Vulkan support! it's the only backend that will make it usable for LLMs on Windows and Linux

**Comment by hicaru** [comment]
Yes please very wait for Vulkan, Rockx. Thanks you guys for your time!
